README.md
Sveltish
================

<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->

`Svelte Stores` are one of the secret weapons of the [Svelte
framework](https://svelte.dev/) (the recently voted [most loved web
framework](https://insights.stackoverflow.com/survey/2021#section-most-loved-dreaded-and-wanted-web-frameworks)).

Stores allow easy [reactive
programming](https://en.wikipedia.org/wiki/Reactive_programming) by
presenting an [Observer
pattern](https://en.wikipedia.org/wiki/Observer_pattern) that is as
simple as necessary, but not simpler.

## Install

``` sh
pip install sveltish
```

## How to use

Sometimes, you’ll have values that need to be accessed by multiple
unrelated objects.

For that, you can use `stores`. It is a very simple implementation
(around 100 lines of code) of the Observer/Observable pattern.

A store is simply an object with a `subscribe` method that allows
interested parties to be notified when its value changes.

#### **Writable Stores**

``` python
from sveltish.stores import writable
```

``` python
count = writable(0)
history = []  # logging for testing
# subscribe returns an unsubscriber
def record(x): 
    history.append(x)
    print(history)
stop = count.subscribe(record)

test_eq(history, [0])
```

    [0]

We just created a `count` store. Its value can be accessed via a
`callback` we pass in the `count.subscribe` method:

A **Writable** can be set from the outside. When it happens, all its
subscribers will react.

``` python
def increment(): count.update(lambda x: x + 1)
def decrement(): count.update(lambda x: x - 1)
def reset(): count.set(0)

count.set(3)
increment()
decrement()
decrement()
reset()
count.set(42)

test_eq(history, [0, 3, 4, 3, 2, 0, 42])
```

    [0, 3]
    [0, 3, 4]
    [0, 3, 4, 3]
    [0, 3, 4, 3, 2]
    [0, 3, 4, 3, 2, 0]
    [0, 3, 4, 3, 2, 0, 42]

The `unsubscriber`, in this example the `stop` function, stops the
notifications to the `subscriber`.

``` python
stop()
reset()
count.set(22)
test_eq(history, [0, 3, 4, 3, 2, 0, 42])
count
```

    w<0> $int: 22

Notice that you can still change the `store` but there was no print
message this time. There was no observer listening.

<div>

> **Note**
>
> [`Observer`](https://fredguth.github.io/sveltish/signals.html#observer),
> `Subscriber` and
> [`Callback`](https://fredguth.github.io/sveltish/signals.html#callback)
> are used as synomyms here.

</div>

When we subscribe new callbacks, they will be promptly informed of the
current state of the `store`.

``` python
stop  = count.subscribe(lambda x: print(f"Count is now {x}"))
stop2 = count.subscribe(lambda x: print(f"double of count is {2*x}"))
```

    Count is now 22
    double of count is 44

``` python
reset()
```

    Count is now 0
    double of count is 0

``` python
stop()
stop2()
```

You can create an empty `Writable Store`.

``` python
store = writable()
history = []
unsubscribe = store.subscribe(lambda x: history.append(x))
unsubscribe()
test_eq(history, [None])
```

If you try to unsubscribe twice, it won’t break. It just does nothing
the second time… and in the third time… and…

``` python
unsubscribe(), unsubscribe(), unsubscribe()
```

    (None, None, None)

Stores assume mutable objects.

<div>

> **Note**
>
> In Python everythong is an object. Here we are calling an object
> something that is not a primitive (eg. int, bool, etc)

</div>

``` python
class Bunch:
    __init__ = lambda self, **kw: setattr(self, '__dict__', kw)

obj = Bunch()
called = 0
store = writable(obj)
def callback(x):
    global called
    called += 1
stop = store.subscribe(callback)
```

``` python
test_eq(called, 1)
obj.a = 1 #type: ignore
store.set(obj)
test_eq(called, 2)
```

#### **Readable Stores**

However… It is clear that not all stores should be writable by whoever
has a reference to them. Many times you want a single `publisher` of
change in store that is only consumed (`subscribed`) by many other
objects. For those cases, we have readable stores.

<div>

> **Note**
>
> The `Publisher Subscriber (PubSub)` pattern is a variant of the
> `Observable/Observer` pattern.

</div>

``` python
from sveltish.stores import readable
```

A Readable store without a `start` function is a constant value and has
no meaning for us. Therefore, `start` is a required argument.

``` python
try:
    c = readable(0) # shoud fail
except Exception as error:
    print(error)

test_fail(lambda: readable(0))
```

    readable() missing 1 required positional argument: 'start'

``` python
class Publisher:
    def __init__(self): self.set = lambda x: None
    def set_set(self, set): 
        self.set = set
        return lambda: None
    def use_set(self, value): self.set(value)
```

``` python
p = Publisher()
reader = readable(0, p.set_set)
reader
```

    r<0> $int: 0

Ths store only starts updating after the first subscriber. Here, the
publisher does not change the store.

``` python
p.use_set(1), reader
```

    (None, r<0> $int: 0)

``` python
stop = reader.subscribe(lambda x: print(f"reader is now {x}"))
```

    reader is now 0

``` python
p.use_set(2)
```

    reader is now 2

``` python
stop()
```

Another example of Readable Store usage:

``` python
from threading import Event, Thread
import time
```

``` python
def start(set): # the start function is the publisher
    stopped = Event()
    def loop(): # needs to be in a separate thread
        while not stopped.wait(1): # in seconds
            set(time.localtime())
    Thread(target=loop).start()    
    return stopped.set
```

``` python
now = readable(time.localtime(), start)
now
```

    r<0> $struct_time: time.struct_time(tm_year=2023, tm_mon=3, tm_mday=8, tm_hour=22, tm_min=12, tm_sec=41, tm_wday=2, tm_yday=67, tm_isdst=0)

<div>

> **Note**
>
> The `loop` needs to be in its own thread, otherwise the function would
> never return and we would wait forever.

</div>

While there is no subscriber, the Readable will not be updated.

``` python
now
```

    r<0> $struct_time: time.struct_time(tm_year=2023, tm_mon=3, tm_mday=8, tm_hour=22, tm_min=12, tm_sec=41, tm_wday=2, tm_yday=67, tm_isdst=0)

``` python
OhPleaseStop = now.subscribe(lambda x: print(time.strftime(f"%H:%M:%S", x), end="\r"))
```

    22:12:41

``` python
time.sleep(2)
OhPleaseStop()
```

    22:12:43

<div>

> **Note**
>
> The Svelte Store api allow you to create a Readable Store without a
> Notifier. See discussion
> [here.](https://github.com/sveltejs/svelte/issues/8300)

</div>

#### **Derived Stores**

A `Derived Store` stores a value based on the value of another store.

``` python
from sveltish.stores import derived
```

For example:

``` python
count = writable(1)
stopCount = count.subscribe(lambda x: print(f"count is {x}"))
double = derived(count, lambda x: x * 2)
stopDouble = double.subscribe(lambda x: print(f"double is {x}"))
test_eq(double.get(), 2*count.get())
```

    count is 1
    double is 2

``` python
count.set(2)
test_eq(double.get(), 4)
```

    count is 2
    double is 4

``` python
stopCount(), stopDouble()
```

    (None, None)

Building on our previous example, we can create a store that derives the
elapsed time since the original store was started.

``` python
elapsing = None
def calc_elapsed(now):
    global elapsing
    if not elapsing: 
        elapsing = now
    return time.mktime(now) - time.mktime(elapsing)
```

``` python
now
```

    r<0> $struct_time: time.struct_time(tm_year=2023, tm_mon=3, tm_mday=8, tm_hour=22, tm_min=12, tm_sec=43, tm_wday=2, tm_yday=67, tm_isdst=0)

``` python
elapsed = derived(now, lambda x: calc_elapsed(x))
elapsed
```

    r<0> $float: 0.0

``` python
stopElapsed = elapsed.subscribe(lambda x: print(f"Elapsed time of source store: {x} seconds.", end="\r"))
```

    Elapsed time of source store: 0.0 seconds.

``` python
time.sleep(1)
stopElapsed()
```

    Elapsed time of source store: 2.0 seconds.

Derived stores allow us to transform the value of a store. In RxPy they
are called `operators`. You can build several operators like: `filter`,
`fold`, `map`, `zip`…

Let’s build a custom `filter` operator:

``` python
user = writable({"name": "John", "age": 32})
stopLog = user.subscribe(lambda x: print(f"User: {x}"))
```

    User: {'name': 'John', 'age': 32}

``` python
name = derived(user, lambda x: x["name"])
stopName = name.subscribe(lambda x: print(f"Name: {x}"))
```

    Name: John

``` python
user.update(lambda x: x | {"age": 45})
```

    User: {'name': 'John', 'age': 45}

Updating the age does not trigger the `name subscriber`. Let’s see what
happens when we update the name.

``` python
user.update(lambda x: x | {"name": "Fred"})
```

    Name: Fred
    User: {'name': 'Fred', 'age': 45}

Only changes to the name of the user triggers the `name` subscriber.

``` python
stopName(), stopLog()
```

    (None, None)

Another cool thing about Derived Stores is that you can derive from a
list of stores. Let’s build a `zip` operator.

``` python
a = writable([1,2,3,4])
b = writable([5,6,7,8])
a,b
```

    (w<0> $list: [1, 2, 3, 4], w<0> $list: [5, 6, 7, 8])

``` python
zipper = derived([a,b], lambda a,b: list(zip(a,b)))
```

``` python
test_eq(zipper.get(), [(1, 5), (2, 6), (3, 7), (4, 8)])
```

While `zipper` has no subscribers, it keeps the initial value, it is
`stopped`.

``` python
a.set([4,3,2,1])
test_eq(zipper.get(), [(1, 5), (2, 6), (3, 7), (4, 8)])
```

A subscription `starts` zipper and it will start to react to the changes
of the stores.

``` python
u = zipper.subscribe(lambda x: None)
test_eq(zipper.get(), [(4, 5), (3, 6), (2, 7), (1, 8)])
```

``` python
b.set([8,7,6,5])
test_eq(zipper.get(), [(4, 8), (3, 7), (2, 6), (1, 5)])
```

``` python
u()
```

#### Store composition with pipes

``` python
writable(1).pipe(lambda x: x + 1).pipe(lambda x: x * 2)
```

    r<0> $int: 4

``` python
writable(1).pipe(lambda x: x+1, lambda x: x*2)
```

    r<0> $int: 4

``` python
writable(1) | (lambda x: x+1) | (lambda x: x*2)
```

    r<0> $int: 4

``` python
a = writable(1)
u5 = (a 
      | (lambda x: x*2) 
      | (lambda x: x*2) 
      | (lambda x: x*2)).subscribe(lambda x: print(f"u5: {x}"))
```

    u5: 8

``` python
a.set(2)
```

    u5: 16

``` python
u5()
```

## Missing features

You may have noticed that along the way we had always to subscribe and
then had to remember to unsubscribe when we were done. This is a bit of
a nuisance. Svelte has a compiler that provide some [syntatic
sugar](https://svelte.dev/tutorial/auto-subscriptions) to make this
easier. They call it `auto-subscriptions`.

`Sveltish` does not have `auto-subscriptions` yet. But if you have a
nice idea how to implement it, please let me know.


CHANGELOG.md
# Release notes

<!-- do not remove -->

## 1.2.7
- feat get and set store value from outside



## 1.2.6
- refactor store representation



## 1.2.4
- add store factories
- Refactors stores


## 1.2.1
- add Pipe operator
- add compose utility



## 1.2.1
- add Pipe operator
- add compose utility



## 1.1.2
- Improves documentation



## 1.0.2
- Initial release


svelty/store.py
# AUTOGENERATED! DO NOT EDIT! File to edit: ../nbs/store.ipynb.

# %% ../nbs/store.ipynb 2
from __future__ import annotations
from typing import Callable, TypeVar,  Generic, Union, Optional, Set, Protocol

# %% auto 0
__all__ = ['T', 'covT', 'Subscriber', 'Unsubscriber', 'Updater', 'Notifier', 'StoreProtocol', 'Writable', 'Readable', 'Derived']

# %% ../nbs/store.ipynb 7
T = TypeVar("T")
covT = TypeVar("covT", covariant=True)
Subscriber = Callable[[T], None] # a callback
Unsubscriber = Callable[[], None] # a callback to be used upon termination of the subscription    
Updater = Callable[[T], T]

# %% ../nbs/store.ipynb 8
class StoreProtocol(Protocol, Generic[covT]):
    ''' The Svelte Store ~~contract~~ protocol. '''
    def subscribe(self, subscriber: Subscriber[T]) -> Unsubscriber: ...

# %% ../nbs/store.ipynb 9
class Store(StoreProtocol[T]):
    ''' A base class for all stores.'''
    value: T
    subscribers: Set[Subscriber]
    def __init__(self, /, **kwargs): 
        self.__dict__.update(kwargs) # see SimpleNamespace: https://docs.python.org/3/library/types.html
    def __repr__(self) -> str:
        return f"{self.__class__.__name__}({self.get()!r})"
    def subscribe(self, callback: Subscriber) -> Unsubscriber:
        return lambda: None
    def get(self) -> T: return self.value

class Readable(Store[T]): pass

class Writable(Store[T]):
    set: Subscriber
    update: Optional[Callable[[Updater],None]] = None

# %% ../nbs/store.ipynb 12
from .util import *

# %% ../nbs/store.ipynb 43
Notifier = Callable[[Subscriber], Union[Unsubscriber, None]]

# %% ../nbs/store.ipynb 44
class Writable(Store[T]):
    ''' A Writable Store.'''
    def __init__(self, 
                 initial_value: T=None, # The initial value of the store
                 start: Notifier=lambda x: None # A Notifier (Optional)
                 ) -> None:
        self.value: T = initial_value
        self.subscribers: Set[Subscriber] = set()
        self.stop: Optional[Unsubscriber] = None
        self.start: Notifier = start

    def subscribe(self, callback: Subscriber) -> Unsubscriber:
        self.subscribers.add(callback)
        if (len(self.subscribers) == 1):
            self.stop = self.start(callback) or (lambda: None)
        callback(self.value)

        def unsubscribe() -> None:
            self.subscribers.remove(callback) if callback in self.subscribers else None
            if (len(self.subscribers) == 0):
                self.stop() if self.stop else None
                self.stop = None
        return unsubscribe
    
    def set(self, new_value: T) -> None:
        if (safe_not_equal(self.value, new_value)):
            self.value = new_value
            for subscriber in self.subscribers:
                subscriber(new_value)
    
    def update(self, fn: Callable[[T], T]) -> None:
        self.set(fn(self.value))
    
    def __len__(self) -> int:
        return len(self.subscribers)

# %% ../nbs/store.ipynb 59
class Readable(Writable[T]):
    ''' A Readable Store.''' 
    def __init__(self, 
                 initial_value: T, # The initial value of the store
                 start: Notifier # A Notifier 
                ) -> None:
        super().__init__(initial_value, start)
    def set(self, *args, **kwargs): raise Exception("Cannot set a Readable Store.")
    def update(self, *args, **kwargs): raise Exception("Cannot update a Readable Store.")

# %% ../nbs/store.ipynb 73
class Derived(Writable):
    ''' A Derived Store.'''
    def __init__(self,
                  source: Store, # The source store
                  fn: Updater # A function that takes the source store's value and returns a new value
                  ) -> None:
        self.target = Writable(source.get())
        self.start: Notifier = lambda x: self.target.set(fn(x))
        self.stop = source.subscribe(self.start)
    def get(self) -> T: return self.target.get()
    def set(self, *args, **kwargs): raise Exception("Cannot set a Derived Store.")
    def update(self, *args, **kwargs): raise Exception("Cannot update a Derived Store.")
    def subscribe(self, callback: Subscriber) -> Unsubscriber:
        return self.target.subscribe(callback)

# %% ../nbs/store.ipynb 95
from fastcore.basics import patch
from .util import retrieve_name


sveltish/util.py
# AUTOGENERATED! DO NOT EDIT! File to edit: ../nbs/10_utils.ipynb.

# %% auto 0
__all__ = ['safe_not_equal', 'compose']

# %% ../nbs/10_utils.ipynb 3
def safe_not_equal(a,b):
    "Check if `a` is not equal to `b`"
    primitive = (int, str, bool, frozenset, tuple)
    return (a != b) if isinstance(a, primitive) else True

# %% ../nbs/10_utils.ipynb 5
from functools import reduce
from typing import Callable, TypeVar,  Generic, Union, Optional, Set, Protocol, Any
from typing_extensions import Annotated

# %% ../nbs/10_utils.ipynb 6
def compose(*functions):
    """Compose multiple functions left to right.

    Examples:
        >>> compose()(x) == x
        >>> compose(f)(x) == f(x)
        >>> compose(f, g)(x) == g(f(x))
        >>> compose(f, g, h)(x) == h(g(f(x)))
        ...
    Returns:
        The composed function.
    """
    if (len(functions)==0): return lambda x: x
    def pack(x): return x if type(x) is tuple else (x,)
    def call(f, g):
       return lambda *x: g(*pack(f(*pack(x))))
    return reduce(call, functions)


sveltish/utils.py
# AUTOGENERATED! DO NOT EDIT! File to edit: ../nbs/10_utils.ipynb.

# %% auto 0
__all__ = ['noop', 'identity', 'safe_not_equal', 'compose', 'Bunch', 'NamedBunch']

# %% ../nbs/10_utils.ipynb 3
def noop(*args, **kwargs): return None

# %% ../nbs/10_utils.ipynb 5
def identity(x): return x

# %% ../nbs/10_utils.ipynb 7
def safe_not_equal(a,b):
    "Check if `a` is not equal to `b`"
    primitive = (int, str, bool, frozenset, tuple)
    return (a != b) if isinstance(a, primitive) else True

# %% ../nbs/10_utils.ipynb 9
from functools import reduce
from typing import Callable, TypeVar,  Generic, Union, Optional, Set, Protocol, Any
from typing_extensions import Annotated

# %% ../nbs/10_utils.ipynb 10
def compose( 
    *functions # functions to be composed (left to right)
) -> Callable[[Any], Any]:  # composed function
    """Compose multiple functions left to right.\n
    Examples:\n
        compose()(x) = x 
        compose(f)(x) = f(x) 
        compose(f, g)(x) = g(f(x)) 
        ... 
    """    
    if (len(functions)==0): return lambda x: x # compose()(x) = x
    def pack(x): return x if type(x) is tuple else (x,)
    def call(f, g):
       return lambda *x: g(*pack(f(*pack(x)))) # call in reverse order
    return reduce(call, functions)  # composed function

# %% ../nbs/10_utils.ipynb 15
class Bunch(dict):
    __init__     = lambda self, **kw: setattr(self, '__dict__', kw) #type: ignore
    __repr__     = lambda self: f'{self.__class__.__name__}({self.__dict__})'
    __contains__ = lambda self, k: k in self.__dict__ or hasattr(self, k)
    __bool__     = lambda self: bool(self.__dict__)
    __hash__     = lambda self: sum([hash(v) if not isinstance(v, (list, set, dict)) else len(v) for v in self.__dict__.values()]) #type: ignore
    asDict       = lambda self: self.__dict__
    asTuple      = lambda self: tuple(self.__dict__.values())

class NamedBunch(Bunch):
    def __init__(self, name, **kw):
        super().__init__(**kw)
        self.__class__.__name__ = name


sveltish/_modidx.py
# Autogenerated by nbdev

d = { 'settings': { 'branch': 'main',
                'doc_baseurl': '/sveltish',
                'doc_host': 'https://fredguth.github.io',
                'git_url': 'https://github.com/fredguth/sveltish',
                'lib_path': 'sveltish'},
  'syms': { 'sveltish.signals': { 'sveltish.signals.Callback': ('signals.html#callback', 'sveltish/signals.py'),
                                  'sveltish.signals.Callback.__init__': ('signals.html#callback.__init__', 'sveltish/signals.py'),
                                  'sveltish.signals.Observable': ('signals.html#observable', 'sveltish/signals.py'),
                                  'sveltish.signals.Observable.subscribe': ('signals.html#observable.subscribe', 'sveltish/signals.py'),
                                  'sveltish.signals.Observer': ('signals.html#observer', 'sveltish/signals.py'),
                                  'sveltish.signals.Observer.cancel': ('signals.html#observer.cancel', 'sveltish/signals.py'),
                                  'sveltish.signals.Observer.run': ('signals.html#observer.run', 'sveltish/signals.py'),
                                  'sveltish.signals.Readable': ('signals.html#readable', 'sveltish/signals.py'),
                                  'sveltish.signals.Readable.__init__': ('signals.html#readable.__init__', 'sveltish/signals.py'),
                                  'sveltish.signals.Readable.__repr__': ('signals.html#readable.__repr__', 'sveltish/signals.py'),
                                  'sveltish.signals.Readable.asTuple': ('signals.html#readable.astuple', 'sveltish/signals.py'),
                                  'sveltish.signals.Signal': ('signals.html#signal', 'sveltish/signals.py'),
                                  'sveltish.signals.Signal.__init__': ('signals.html#signal.__init__', 'sveltish/signals.py'),
                                  'sveltish.signals.Writable': ('signals.html#writable', 'sveltish/signals.py'),
                                  'sveltish.signals.Writable.__init__': ('signals.html#writable.__init__', 'sveltish/signals.py'),
                                  'sveltish.signals.computed': ('signals.html#computed', 'sveltish/signals.py'),
                                  'sveltish.signals.pipe': ('signals.html#pipe', 'sveltish/signals.py'),
                                  'sveltish.signals.reaction': ('signals.html#reaction', 'sveltish/signals.py'),
                                  'sveltish.signals.readonly': ('signals.html#readonly', 'sveltish/signals.py'),
                                  'sveltish.signals.signal': ('signals.html#signal', 'sveltish/signals.py'),
                                  'sveltish.signals.writable': ('signals.html#writable', 'sveltish/signals.py')},
            'sveltish.stores': { 'sveltish.stores.DerivedStore': ('stores.html#derivedstore', 'sveltish/stores.py'),
                                 'sveltish.stores.DerivedStore.__init__': ('stores.html#derivedstore.__init__', 'sveltish/stores.py'),
                                 'sveltish.stores.DerivedStore.get': ('stores.html#derivedstore.get', 'sveltish/stores.py'),
                                 'sveltish.stores.DerivedStore.set': ('stores.html#derivedstore.set', 'sveltish/stores.py'),
                                 'sveltish.stores.DerivedStore.subscribe': ('stores.html#derivedstore.subscribe', 'sveltish/stores.py'),
                                 'sveltish.stores.DerivedStore.update': ('stores.html#derivedstore.update', 'sveltish/stores.py'),
                                 'sveltish.stores.ReadableStore': ('stores.html#readablestore', 'sveltish/stores.py'),
                                 'sveltish.stores.ReadableStore.__init__': ('stores.html#readablestore.__init__', 'sveltish/stores.py'),
                                 'sveltish.stores.ReadableStore.__repr__': ('stores.html#readablestore.__repr__', 'sveltish/stores.py'),
                                 'sveltish.stores.ReadableStore.set': ('stores.html#readablestore.set', 'sveltish/stores.py'),
                                 'sveltish.stores.ReadableStore.update': ('stores.html#readablestore.update', 'sveltish/stores.py'),
                                 'sveltish.stores.Store': ('stores.html#store', 'sveltish/stores.py'),
                                 'sveltish.stores.Store.__getattr__': ('stores.html#store.__getattr__', 'sveltish/stores.py'),
                                 'sveltish.stores.Store.__init__': ('stores.html#store.__init__', 'sveltish/stores.py'),
                                 'sveltish.stores.Store.__len__': ('stores.html#store.__len__', 'sveltish/stores.py'),
                                 'sveltish.stores.Store.__repr__': ('stores.html#store.__repr__', 'sveltish/stores.py'),
                                 'sveltish.stores.Store.__set': ('stores.html#store.__set', 'sveltish/stores.py'),
                                 'sveltish.stores.Store.__setattr__': ('stores.html#store.__setattr__', 'sveltish/stores.py'),
                                 'sveltish.stores.Store.get': ('stores.html#store.get', 'sveltish/stores.py'),
                                 'sveltish.stores.Store.set': ('stores.html#store.set', 'sveltish/stores.py'),
                                 'sveltish.stores.Store.subscribe': ('stores.html#store.subscribe', 'sveltish/stores.py'),
                                 'sveltish.stores.Store.update': ('stores.html#store.update', 'sveltish/stores.py'),
                                 'sveltish.stores.StoreProtocol': ('stores.html#storeprotocol', 'sveltish/stores.py'),
                                 'sveltish.stores.StoreProtocol.subscribe': ('stores.html#storeprotocol.subscribe', 'sveltish/stores.py'),
                                 'sveltish.stores.Writable': ('stores.html#writable', 'sveltish/stores.py'),
                                 'sveltish.stores.Writable.set': ('stores.html#writable.set', 'sveltish/stores.py'),
                                 'sveltish.stores.Writable.update': ('stores.html#writable.update', 'sveltish/stores.py'),
                                 'sveltish.stores.__or__': ('stores.html#__or__', 'sveltish/stores.py'),
                                 'sveltish.stores.derived': ('stores.html#derived', 'sveltish/stores.py'),
                                 'sveltish.stores.pipe': ('stores.html#pipe', 'sveltish/stores.py'),
                                 'sveltish.stores.readable': ('stores.html#readable', 'sveltish/stores.py'),
                                 'sveltish.stores.writable': ('stores.html#writable', 'sveltish/stores.py')},
            'sveltish.utils': { 'sveltish.utils.Bunch': ('utils.html#bunch', 'sveltish/utils.py'),
                                'sveltish.utils.NamedBunch': ('utils.html#namedbunch', 'sveltish/utils.py'),
                                'sveltish.utils.NamedBunch.__init__': ('utils.html#namedbunch.__init__', 'sveltish/utils.py'),
                                'sveltish.utils.compose': ('utils.html#compose', 'sveltish/utils.py'),
                                'sveltish.utils.identity': ('utils.html#identity', 'sveltish/utils.py'),
                                'sveltish.utils.noop': ('utils.html#noop', 'sveltish/utils.py'),
                                'sveltish.utils.safe_not_equal': ('utils.html#safe_not_equal', 'sveltish/utils.py')}}}


sveltish/stores.py
# AUTOGENERATED! DO NOT EDIT! File to edit: ../nbs/00_stores.ipynb.

# %% ../nbs/00_stores.ipynb 0
from __future__ import annotations
from typing import Callable, TypeVar,  Generic, Union, Optional, Set, Protocol, Any, Sequence
from fastcore.test import test_eq, test_fail

# %% auto 0
__all__ = ['T', 'covT', 'Subscriber', 'Unsubscriber', 'Updater', 'Notifier', 'Readable', 'StoreProtocol', 'Writable', 'Store',
           'writable', 'ReadableStore', 'readable', 'DerivedStore', 'derived', 'pipe']

# %% ../nbs/00_stores.ipynb 6
T = TypeVar("T")
covT = TypeVar("covT", covariant=True)
Subscriber = Callable[[T], None] # a callback
Unsubscriber = Callable[[], None] # a callback to be used upon termination of the subscription
Updater = Callable[[T], T]
Notifier = Callable[[Subscriber], Union[Unsubscriber, None]]

class StoreProtocol(Protocol, Generic[covT]):
    ''' The Svelte Store ~~contract~~ protocol. '''
    def subscribe(self, subscriber: Subscriber[T]) -> Unsubscriber: ...

Readable: TypeAlias = StoreProtocol[T]

class Writable(Readable[T]):
    ''' Writable protocol'''
    def set(self, value: T) -> None: ...
    def update(self, updater: Updater[T]) -> None: ...

# %% ../nbs/00_stores.ipynb 9
import sveltish.utils as utils

# %% ../nbs/00_stores.ipynb 10
class Store(Readable[T]):
    ''' A Writable Store.'''
    def __init__(self:Writable,
                initial_value: Any = None, # initial value of the store
                start: Notifier = utils.noop # A Notifier (Optional)
                ) -> None:
        self.value = initial_value
        self.subscribers: Set[Subscriber] = set() # callbacks to be called when the value changes
        self.start: Notifier = start # function called when the first subscriber is added
        self.stop: Optional[Unsubscriber] = None  # functional called when the last subscriber is removed

    def get(self) -> T: return self.value
    __call__ = get

    def subscribe(self:Writable,
                  callback: Subscriber # callback to be called when the store value changes
                  ) -> Unsubscriber:
        ''' Adds callback to the list of subscribers.'''
        self.subscribers.add(callback)
        if (len(self.subscribers) == 1):
            self.stop = self.start(self.__set) or (lambda: None) #type: ignore
        callback(self.value)
        def unsubscribe() -> None:
            ''' Removes callback from the list of subscribers.'''
            self.subscribers.remove(callback) if callback in self.subscribers else None
            if (len(self.subscribers) == 0):
                self.stop() if self.stop else None #type: ignore
                self.stop = None #type: ignore
        return unsubscribe

    def __set(self,
            new_value: T # The new value of the store
            ) -> None:
        ''' Internal implementation of set used inside Readable Store, which does not exposes set.'''
        if (utils.safe_not_equal(self.value, new_value)):
            self.value = new_value
            for subscriber in self.subscribers:
                subscriber(new_value)

    def set(self,
            new_value: T # The new value of the store
            ) -> None:
        ''' Set value of store.'''
        self.__set(new_value)

    def update(self,
               fn: Callable[[T], T] # a callback that takes the existing store value and updates it
               ) -> None:
        ''' Update the store value by applying `fn` to the existing value.'''
        self.set(fn(self.value))

    def __len__(self) -> int:
        ''' The length of the store is the number of subscribers.'''
        return len(self.subscribers)

    def __repr__(self) -> str:
        return f"w<{len(self)}> ${self.value.__class__.__name__}: {self.value.__repr__()}"

    def __getattr__(self, k):
        '''Called if property not found in Store object.'''
        conditions = [
            isinstance(k, (int, str)),
            not isinstance(self, DerivedStore), #type: ignore
            k[:2]!='__'
        ]
        if all(conditions):
            if isinstance(self.value, dict) and k in self.value:
                return self.value[k] # look in Store value
            if hasattr(self.value, k): 
                return getattr(self.value, k) # look in Store value
        raise AttributeError(k)

    def __setattr__(self, k:str,v) -> None:
        exceptions = ['value', 'subscribers', 'start', 'stop', 'sources', 'fn', 'target']

        if k not in exceptions:
            new_value = self.value
            if isinstance(self.value, dict) and k in self.value:
                new_value = {**self.value, k:v}
            if hasattr(self.value, k):
                new_value = self.value.__class__(**{**self.value.__dict__, k:v})
            # uses set instead of __set because this shouldn't work with readable store 
            self.set(new_value)
        else: super().__setattr__(k,v)

# %% ../nbs/00_stores.ipynb 12
def writable(value: T = None, # initial value of the store
             start: Notifier = utils.noop # Optional Notifier, a function called when the first subscriber is added
             ) -> Writable[T]: # Writable Store
    ''' Creates a new Writable Store (A Writable factory).'''
    return Store(value, start)

# %% ../nbs/00_stores.ipynb 15
class ReadableStore(Store[T]):
    ''' A Readable Store.'''
    def __init__(self,
                 initial_value: T, # initial value of the store
                 start: Notifier # function called when the first subscriber is added
                ) -> None:
        super().__init__(initial_value, start)
    def set(self, *args, **kwargs): raise Exception("Cannot set a Readable Store.")
    def update(self, *args, **kwargs): raise Exception("Cannot update a Readable Store.")
    def __repr__(self) -> str: return "r" + super().__repr__()[1:]

# %% ../nbs/00_stores.ipynb 17
def readable(value: T, # initial value of the store
             start: Notifier  # function called when the first subscriber is added
             ) -> Readable[T]:  # Readable Store
    ''' Creates a new Readable Store (A Readable factory).'''
    return ReadableStore(value, start)

# %% ../nbs/00_stores.ipynb 21
from .utils import compose
from fastcore.foundation import L

# %% ../nbs/00_stores.ipynb 22
class DerivedStore(Store[T]):
    ''' A Derived Store.'''
    def __init__(self,
                 s: Union[Store, list[Store]], # source store(s)
                 *functions: Callable, # a callback that takes the source store(s) values and returns the derived value
             ) -> None:
        self.sources = L(s)
        if not all(isinstance(x, Store) for x in self.sources):
            raise Exception("s must be a Store or a list of Stores")
        self.fn = compose(*functions)

        def start(set_fn: Subscriber):
            def sync(x=None): # x is ignored
                values = self.sources.map(lambda x: x.get())
                set_fn(self.fn(*values))
            sync() # sync target with source values, they can have changed since Derived creation
            unsubscribers = self.sources.map(lambda s: s.subscribe(sync))
            def stop():
                for unsubscribe in unsubscribers: unsubscribe()
            return stop
        values = self.sources.map(lambda x: x.get())
        self.target = readable(self.fn(*values), start)

    def get(self): return self.target.get()
    def set(self, *args, **kwargs): raise Exception("Cannot set a Derived Store.")
    def update(self, *args, **kwargs): raise Exception("Cannot update a Derived Store.")
    def subscribe(self,
                  callback: Subscriber # callback to be called when any of the source stores change
                  ) -> Unsubscriber:
        ''' Adds callback to the list of subscribers.'''
        return self.target.subscribe(callback)

# %% ../nbs/00_stores.ipynb 24
def derived(s: Union[Store, list[Store]], # source store(s)
            *functions: list(Callable[...,T]) # a callback that takes the source store(s) values and returns the derived value
            ) -> Readable: # Derived Store
    ''' Creates a new Derived Store (A Derived factory).'''
    return DerivedStore(s, *functions).target

# %% ../nbs/00_stores.ipynb 28
import fastcore.all as fc

# %% ../nbs/00_stores.ipynb 29
@fc.patch
def pipe(self:Store, # source store
         *functions: list(Callable[...,T]) # functions that transform the source store
         )->Readable[T]: # returned store
     ''' Unix-like Pipe operator.'''
     return derived(self, *functions)

# %% ../nbs/00_stores.ipynb 31
@fc.patch
def __or__(self:Store, # source store
           other: Callable[...,T] # function that transforms the source store
           ) -> Readable[T]: # returned store
    ''' self | other  works like Unix pipes. It returns a Derived Store that is the result of applying other to self.'''
    return self.pipe(other)
